internal/authentication/ldap_client_factory.go

```diff
@@ -117,162 +121,393 @@ type PooledLDAPClientFactory struct {
+			f.config.Pooling.Timeout, minRecommendedTimeout)
+	}
 
-	for i := 0; i < f.config.Pooling.Count; i++ {
-		if client, err = f.new(); err != nil {
-			errs = append(errs, err)
+	f.pool = make(chan *LDAPClientPooled, f.maxPoolSize)
+	f.wakeup = make(chan struct{}, 1)
+	f.closed = make(chan error, 1)
 
-			continue
-		}
-
-		f.pool <- client
-	}
+	f.ctx, f.cancel = context.WithCancel(context.Background())
 
-	if len(errs) == f.config.Pooling.Count {
-		return fmt.Errorf("errors occurred initializing the client pool: no connections could be established")
+	// prime the pool to test connectivity and run a quick self-check
+	if !f.tryAddPooledClient(f.pool) {
+		err = fmt.Errorf("LDAP pool couldn't acquire initial client")
+	} else if rerr := f.ReadinessCheck(); rerr != nil {
+		err = rerr // failed readiness check
+	} else {
+		go f.poolManager() // transfer responsibility to the pool manager
+		return nil
 	}
 
-	return nil
+	logging.Logger().Debug("LDAP pool initialization failed, aborting...")
+	err = fmt.Errorf("LDAP pool initialization failed: %w", err)
+	f.closed<-err
+	_ = f.Close()
+	return err
 }
 
-// GetClient opens new client using the pool.
+// thread-safe, opens new client using the pool.
 func (f *PooledLDAPClientFactory) GetClient(opts ...LDAPClientFactoryOption) (conn ldap.Client, err error) {
 	if len(opts) != 0 {
 		return getLDAPClient(f.config.Address.String(), f.config.User, f.config.Password, f.config.Timeout, f.dialer, f.tls, f.config.StartTLS, f.opts, opts...)
 	}
 
-	return f.acquire(context.Background())
+	return f.acquire()
 }
 
+// TODO: Clarify why this method is not thread-safe - or does this only apply to the returnd client?
+
 // The new function creates a pool based client. This function is not thread safe.
+// NOTE: this function adheres to config.Timeout, not config.Pooling.Timeout!
 func (f *PooledLDAPClientFactory) new() (pooled *LDAPClientPooled, err error) {
-	var client ldap.Client
-
-	if client, err = getLDAPClient(f.config.Address.String(), f.config.User, f.config.Password, f.config.Timeout, f.dialer, f.tls, f.config.StartTLS, f.opts); err != nil {
-		return nil, fmt.Errorf("error occurred establishing new client for the pool: %w", err)
-	}
+	client, err := getLDAPClient(f.config.Address.String(), f.config.User, f.config.Password, f.config.Timeout, f.dialer, f.tls, f.config.StartTLS, f.opts)
+	if err != nil { return nil, err }
 
 	return &LDAPClientPooled{Client: client}, nil
 }
 
-// ReleaseClient returns a client using the pool or closes it.
+// thread-safe, returns a client using the pool or closes it.
 func (f *PooledLDAPClientFactory) ReleaseClient(client ldap.Client) (err error) {
-	f.mu.Lock()
-
-	defer f.mu.Unlock()
+	logging.Logger().Trace("Releasing LDAP client")
 
-	if f.closing {
+	pooled, ok := client.(*LDAPClientPooled)
+	if !ok {
+		logging.Logger().Debug("LDAP client is not pooled, closing directly")
```

Comment:
Please remove call to logging.Logger() and initialize a fielded logger like in #10383.